sound/soc/soc-core.c

```diff
@@ -2135,13 +2135,17 @@ static int snd_soc_instantiate_card(struct snd_soc_card *card)
 	if (ret < 0) {
 		dev_err(card->dev, "ASoC: failed to register soundcard %d\n",
 				ret);
-		goto probe_end;
+		goto comp_remove;
 	}
 
 	card->instantiated = 1;
 	dapm_mark_endpoints_dirty(card);
 	snd_soc_dapm_sync(&card->dapm);
 
+comp_remove:
+	if (ret < 0)
+		soc_remove_link_components(card);
```

Comment:
If you see @morimoto 's patches and my WIP fix, the remove_link_components() is already called in the cleanup_card_resources (maybe in the wrong location), so wondering if all these changes are actually needed